File: snippets/fsharp/System/Byte/Overview/tobyte3.fs
module tobyte3

// <Snippet12>
open System
open System.Globalization

type SignBit =
   | Negative = -1
   | Zero = 0
   | Positive = 1

[<Struct>]
type ByteString =
    val mutable private value: string
    
    val mutable Sign : SignBit
    member this.Value
        with get () = this.value
        and set (value: string) =
            if value.Trim().Length > 2 then
                invalidArg "value" "The string representation of a byte cannot have more than two characters"
            else
                this.value <- value 

    // IConvertible implementations.
    interface IConvertible with
        member _.GetTypeCode() =
            TypeCode.Object

        member this.ToBoolean(provider: IFormatProvider) =
            this.Sign <> SignBit.Zero

        member this.ToByte(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                raise (OverflowException $"{Convert.ToSByte(this.value, 16)} is out of range of the Byte type.")
            else
                Byte.Parse(this.value, NumberStyles.HexNumber)

        member this.ToChar(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                raise (OverflowException $"{Convert.ToSByte(this.value, 16)} is out of range of the Char type.")
            else
                let byteValue = Byte.Parse(this.value, NumberStyles.HexNumber)
                Convert.ToChar byteValue

        member _.ToDateTime(provider: IFormatProvider) =
            raise (InvalidCastException "ByteString to DateTime conversion is not supported.")

        member this.ToDecimal(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                let byteValue = SByte.Parse(this.Value, NumberStyles.HexNumber)
                Convert.ToDecimal byteValue
            else
                let byteValue = Byte.Parse(this.Value, NumberStyles.HexNumber)
                Convert.ToDecimal byteValue

        member this.ToDouble(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                Convert.ToDouble(SByte.Parse(this.value, NumberStyles.HexNumber))
            else
                Convert.ToDouble(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToInt16(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                Convert.ToInt16(SByte.Parse(this.value, NumberStyles.HexNumber))
            else
                Convert.ToInt16(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToInt32(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                Convert.ToInt32(SByte.Parse(this.value, NumberStyles.HexNumber))
            else
                Convert.ToInt32(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToInt64(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                Convert.ToInt64(SByte.Parse(this.value, NumberStyles.HexNumber))
            else
                Convert.ToInt64(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToSByte(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                try
                    Convert.ToSByte(Byte.Parse(this.value, NumberStyles.HexNumber))
                with :? OverflowException as e ->
                    raise (OverflowException($"{Byte.Parse(this.value, NumberStyles.HexNumber)} is outside the range of the SByte type.", e))
            else
                SByte.Parse(this.value, NumberStyles.HexNumber)

        member this.ToSingle(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                Convert.ToSingle(SByte.Parse(this.value, NumberStyles.HexNumber))
            else
                Convert.ToSingle(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToString(provider: IFormatProvider) =
            "0x" + this.value

        member this.ToType(conversionType: Type, provider: IFormatProvider): obj =
            let this = this :> IConvertible
            match Type.GetTypeCode conversionType with
            | TypeCode.Boolean ->
                this.ToBoolean null
            | TypeCode.Byte ->
                this.ToByte null
            | TypeCode.Char ->
                this.ToChar null
            | TypeCode.DateTime ->
                this.ToDateTime null
            | TypeCode.Decimal ->
                this.ToDecimal null
            | TypeCode.Double ->
                this.ToDouble null
            | TypeCode.Int16 ->
                this.ToInt16 null
            | TypeCode.Int32 ->
                this.ToInt32 null
            | TypeCode.Int64 ->
                this.ToInt64 null
            | TypeCode.Object ->
                if typeof<ByteString>.Equals conversionType then
                    this
                else
                    raise (InvalidCastException $"Conversion to a {conversionType.Name} is not supported.")
            | TypeCode.SByte ->
                this.ToSByte null
            | TypeCode.Single ->
                this.ToSingle null
            | TypeCode.String ->
                this.ToString null
            | TypeCode.UInt16 ->
                this.ToUInt16 null
            | TypeCode.UInt32 ->
                this.ToUInt32 null
            | TypeCode.UInt64 ->
                this.ToUInt64 null
            | _ ->
                raise (InvalidCastException $"Conversion to {conversionType.Name} is not supported.")

        member this.ToUInt16(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                raise (OverflowException $"{SByte.Parse(this.value, NumberStyles.HexNumber)} is outside the range of the UInt16 type.")
            else
                Convert.ToUInt16(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToUInt32(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                raise (OverflowException $"{SByte.Parse(this.value, NumberStyles.HexNumber)} is outside the range of the UInt32 type.")
            else
                Convert.ToUInt32(Byte.Parse(this.value, NumberStyles.HexNumber))

        member this.ToUInt64(provider: IFormatProvider) =
            if this.Sign = SignBit.Negative then
                raise (OverflowException $"{SByte.Parse(this.value, NumberStyles.HexNumber)} is outside the range of the UInt64 type.")
            else
                Convert.ToUInt64(Byte.Parse(this.value, NumberStyles.HexNumber))
// </Snippet12>

// <Snippet13>
let positiveByte = 216uy
let negativeByte = -101y

let mutable positiveString = ByteString()
positiveString.Sign <- Math.Sign positiveByte |> enum
positiveString.Value <- positiveByte.ToString "X2"

let mutable negativeString = ByteString()
negativeString.Sign <- Math.Sign negativeByte |> enum
negativeString.Value <- negativeByte.ToString "X2"

try
    printfn $"'{positiveString.Value}' converts to {Convert.ToByte positiveString}."
with :? OverflowException ->
    printfn $"0x{positiveString.Value} is outside the range of the Byte type."

try
    printfn $"'{negativeString.Value}' converts to {Convert.ToByte negativeString}."
with :? OverflowException ->
    printfn $"0x{negativeString.Value} is outside the range of the Byte type."

// The example displays the following output:
//       'D8' converts to 216.
//       0x9B is outside the range of the Byte type.
// </Snippet13>

File: snippets/csharp/System/Byte/Overview/tobyte3.cs
// <Snippet12>
using System;
using System.Globalization;

public enum SignBit { Negative = -1, Zero = 0, Positive = 1 };

public struct ByteString3 : IConvertible
{
    private SignBit signBit;
    private string byteString;

    public SignBit Sign
    {
        set => signBit = value;
        get => signBit;
    }

    public string Value
    {
        set
        {
            if (value.Trim().Length > 2)
            {
                throw new ArgumentException("The string representation of a byte cannot have more than two characters.");
            }
            else
            {
                byteString = value;
            }
        }
        get => byteString;
    }

    // IConvertible implementations.
    public TypeCode GetTypeCode() => TypeCode.Object;

    public bool ToBoolean(IFormatProvider provider)
    {
        if (signBit == SignBit.Zero)
        {
            return false;
        }
        else
        {
            return true;
        }
    }

    public byte ToByte(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            throw new OverflowException($"{Convert.ToSByte(byteString, 16)} is out of range of the Byte type.");
        }
        else
        {
            return byte.Parse(byteString, NumberStyles.HexNumber);
        }
    }

    public char ToChar(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            throw new OverflowException($"{Convert.ToSByte(byteString, 16)} is out of range of the Char type.");
        }
        else
        {
            byte byteValue = byte.Parse(this.byteString, NumberStyles.HexNumber);
            return Convert.ToChar(byteValue);
        }
    }

    public DateTime ToDateTime(IFormatProvider provider) => throw new InvalidCastException("ByteString3 to DateTime conversion is not supported.");

    public decimal ToDecimal(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            sbyte byteValue = sbyte.Parse(byteString, NumberStyles.HexNumber);
            return Convert.ToDecimal(byteValue);
        }
        else
        {
            byte byteValue = byte.Parse(byteString, NumberStyles.HexNumber);
            return Convert.ToDecimal(byteValue);
        }
    }

    public double ToDouble(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            return Convert.ToDouble(sbyte.Parse(byteString, NumberStyles.HexNumber));
        }
        else
        {
            return Convert.ToDouble(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public short ToInt16(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            return Convert.ToInt16(sbyte.Parse(byteString, NumberStyles.HexNumber));
        }
        else
        {
            return Convert.ToInt16(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public int ToInt32(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            return Convert.ToInt32(sbyte.Parse(byteString, NumberStyles.HexNumber));
        }
        else
        {
            return Convert.ToInt32(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public long ToInt64(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            return Convert.ToInt64(sbyte.Parse(byteString, NumberStyles.HexNumber));
        }
        else
        {
            return Convert.ToInt64(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public sbyte ToSByte(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            try
            {
                return Convert.ToSByte(byte.Parse(byteString, NumberStyles.HexNumber));
            }
            catch (OverflowException e)
            {
                throw new OverflowException($"{byte.Parse(byteString, NumberStyles.HexNumber)} is outside the range of the SByte type.", e);
            }
        }
        else
        {
            return sbyte.Parse(byteString, NumberStyles.HexNumber);
        }
    }

    public float ToSingle(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            return Convert.ToSingle(sbyte.Parse(byteString, NumberStyles.HexNumber));
        }
        else
        {
            return Convert.ToSingle(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public string ToString(IFormatProvider provider) => "0x" + this.byteString;

    public object ToType(Type conversionType, IFormatProvider provider)
    {
        switch (Type.GetTypeCode(conversionType))
        {
            case TypeCode.Boolean:
                return this.ToBoolean(null);
            case TypeCode.Byte:
                return this.ToByte(null);
            case TypeCode.Char:
                return this.ToChar(null);
            case TypeCode.DateTime:
                return this.ToDateTime(null);
            case TypeCode.Decimal:
                return this.ToDecimal(null);
            case TypeCode.Double:
                return this.ToDouble(null);
            case TypeCode.Int16:
                return this.ToInt16(null);
            case TypeCode.Int32:
                return this.ToInt32(null);
            case TypeCode.Int64:
                return this.ToInt64(null);
            case TypeCode.Object:
                if (typeof(ByteString3).Equals(conversionType))
                {
                    return this;
                }
                else
                {
                    throw new InvalidCastException($"Conversion to a {conversionType.Name} is not supported.");
                }

            case TypeCode.SByte:
                return this.ToSByte(null);
            case TypeCode.Single:
                return this.ToSingle(null);
            case TypeCode.String:
                return this.ToString(null);
            case TypeCode.UInt16:
                return this.ToUInt16(null);
            case TypeCode.UInt32:
                return this.ToUInt32(null);
            case TypeCode.UInt64:
                return this.ToUInt64(null);
            default:
                throw new InvalidCastException($"Conversion to {conversionType.Name} is not supported.");
        }
    }

    public ushort ToUInt16(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            throw new OverflowException($"{sbyte.Parse(byteString, NumberStyles.HexNumber)} is outside the range of the UInt16 type.");
        }
        else
        {
            return Convert.ToUInt16(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public uint ToUInt32(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            throw new OverflowException($"{sbyte.Parse(byteString, NumberStyles.HexNumber)} is outside the range of the UInt32 type.");
        }
        else
        {
            return Convert.ToUInt32(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }

    public ulong ToUInt64(IFormatProvider provider)
    {
        if (signBit == SignBit.Negative)
        {
            throw new OverflowException($"{sbyte.Parse(byteString, NumberStyles.HexNumber)} is outside the range of the UInt64 type.");
        }
        else
        {
            return Convert.ToUInt64(byte.Parse(byteString, NumberStyles.HexNumber));
        }
    }
}
// </Snippet12>

// <Snippet13>
public class Class1
{
    public static void Main()
    {
        byte positiveByte = 216;
        sbyte negativeByte = -101;

        ByteString3 positiveString = new()
        {
            Sign = (SignBit)Math.Sign(positiveByte),
            Value = positiveByte.ToString("X2")
        };

        ByteString3 negativeString = new()
        {
            Sign = (SignBit)Math.Sign(negativeByte),
            Value = negativeByte.ToString("X2")
        };

        try
        {
            Console.WriteLine($"'{positiveString.Value}' converts to {Convert.ToByte(positiveString)}.");
        }
        catch (OverflowException)
        {
            Console.WriteLine($"0x{positiveString.Value} is outside the range of the Byte type.");
        }

        try
        {
            Console.WriteLine($"'{negativeString.Value}' converts to {Convert.ToByte(negativeString)}.");
        }
        catch (OverflowException)
        {
            Console.WriteLine($"0x{negativeString.Value} is outside the range of the Byte type.");
        }
    }
}
// The example displays the following output:
//       'D8' converts to 216.
//       0x9B is outside the range of the Byte type.
// </Snippet13>
